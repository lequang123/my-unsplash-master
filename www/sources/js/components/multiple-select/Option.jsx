import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Option extends Component {
    static propTypes = {
        itemData: PropTypes.object.isRequired,
        id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
        onChange: PropTypes.func
    };

    constructor(props) {
        super(props);

        this.state = {
            isExpandedSubData: true
        };
    }

    onChange = event => {
        event.stopPropagation();

        const { itemData, onChange } = this.props;

        const newItemData = {
            key: itemData.key,
            value: itemData.value,
            checked: !itemData.checked
        };
        onChange && onChange(newItemData);
    };

    onClickExpandIcon = () => {
        this.setState(prevState => ({
            isExpandedSubData: !prevState.isExpandedSubData
        }));
    };

    renderSubData = subDataList => {
        return (
            <ul className='sub-data'>
                {
                    this.state.isExpandedSubData &&
                    subDataList.map((item, key) => <li key={`sub-${key}`}>{item}</li>)
                }
            </ul>
        );
    };

    render() {
        const { itemData, id } = this.props;
        const isHasSubData = itemData.subData && itemData.subData.length > 0;
        return (
            <li className="multiple-select-item">
                <input
                    id={id}
                    type="checkbox"
                    className="option-checkbox"
                    checked={itemData.checked}
                    onChange={this.onChange}
                />
                <label className="option-label" htmlFor={id}>
                    {itemData.value}
                </label>
                {
                    isHasSubData &&
                    <span className={`expand-icon fa ${this.state.isExpandedSubData ? 'fa-caret-down' : 'fa-caret-right'}`} onClick={this.onClickExpandIcon} />
                }
                {
                    isHasSubData &&
                    this.renderSubData(itemData.subData)
                }
            </li>
        );
    }
}