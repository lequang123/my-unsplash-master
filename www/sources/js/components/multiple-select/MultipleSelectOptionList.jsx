import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Option from './Option';

export default class MultipleSelectOptionList extends Component {
    static propTypes = {
        dataSource: PropTypes.arrayOf(PropTypes.object).isRequired,
        id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
        onChange: PropTypes.func
    };

    constructor(props) {
        super(props);
    }

    _renderOptionList() {
        const { dataSource, onChange, id } = this.props;
        if (!dataSource || !dataSource.length) {
            return <span>No information</span>;
        }

        const optionList = dataSource.map((item, index) => {
            return item.visible && (
                <Option
                    id={`${id}-optionItem-${index}`}
                    key={`${id}optionItem${index}`}
                    itemData={item}
                    onChange={onChange}
                />
            );
        });

        return optionList;
    }

    _calculateMinHeightOptionList = maxLengthOfSubData => {
        let minHeight = 0;
        if (maxLengthOfSubData && maxLengthOfSubData > 0 && maxLengthOfSubData < 5) {
            minHeight = maxLengthOfSubData * 40;
        }

        if (maxLengthOfSubData > 5) {
            minHeight = 150;
        }

        return minHeight;
    };

    render() {
        const dataSource = this.props.dataSource;
        const maxLengthOfSubData = dataSource.length > 0 ? Math.max(...dataSource.filter(item => item.subData !== null).map(i => i.subData.length)) : 0;
        const minHeight = this._calculateMinHeightOptionList(maxLengthOfSubData);
        return (
            <ul className="multiple-select-options" style={{ minHeight: minHeight }}>
                {this._renderOptionList()}
            </ul>
        );
    }
}