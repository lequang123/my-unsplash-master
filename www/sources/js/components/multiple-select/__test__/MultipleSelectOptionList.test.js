import React from 'react';
import MultipleSelectOptionList from '../MultipleSelectOptionList';
import Option from '../Option';
import { shallow } from 'enzyme';

describe('MultipleSelectOptionList', () => {
    it('should have correct snapshots', () => {
        // Arrange
        const props = {
            dataSource: []
        };

        // Action
        const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);

        // Assert
        expect(multipleSelectOptionList).toMatchSnapshot();
    });

    describe('_renderOptionList', () => {
        it('should return no data when dataSource empty', () => {
            // Arrange
            const props = {
                dataSource: []
            };

            const expected = <span>No information</span>;
            const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);

            // Action
            const result = multipleSelectOptionList.instance()._renderOptionList();

            // Assert
            expect(result).toEqual(expected);
        });
        
        it('should return expected when dataSource not empty', () => {
            // Arrange
            const item = {
                visible: true,
                subData: []
            };

            const props = {
                id: 'id',
                dataSource: [
                    item
                ],
                onChange: jest.fn()
            };

            const expected = [
                <Option
                    id='id-optionItem-0'
                    key='idoptionItem0'
                    itemData={item}
                    onChange={props.onChange}
                />
            ];
            const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);

            // Action
            const result = multipleSelectOptionList.instance()._renderOptionList();

            // Assert
            expect(result).toEqual(expected);
        });
    });

    describe('_calculateMinHeightOptionList', () => {
        it('should return expected when maxLengthOfSubData from 1 to 4', () => {
            // Arrange
            const props = {
                dataSource: []
            };

            const maxLengthOfSubData = 1;
            const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);

            // Action
            const result = multipleSelectOptionList.instance()._calculateMinHeightOptionList(maxLengthOfSubData);

            // Assert
            expect(result).toEqual(maxLengthOfSubData * 40);
        });
        
        it('should return expected when maxLengthOfSubData greater than 5', () => {
            // Arrange
            const props = {
                dataSource: []
            };

            const maxLengthOfSubData = 6;
            const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);

            // Action
            const result = multipleSelectOptionList.instance()._calculateMinHeightOptionList(maxLengthOfSubData);

            // Assert
            expect(result).toEqual(150);
        });
        
        it('should return expected when maxLengthOfSubData equal 5 or not greater than 0', () => {
            // Arrange
            const props = {
                dataSource: []
            };

            const maxLengthOfSubData = 5;
            const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);

            // Action
            const result = multipleSelectOptionList.instance()._calculateMinHeightOptionList(maxLengthOfSubData);

            // Assert
            expect(result).toEqual(0);
        });
    });

    it('render should return correct components', () => {
        // Arrange
        const props = {
            dataSource: []
        };

        const minHeight = 10;
        const _renderOptionList = jest.fn();
        const multipleSelectOptionList = shallow(<MultipleSelectOptionList {...props} />);
        multipleSelectOptionList.instance()._renderOptionList = _renderOptionList;
        multipleSelectOptionList.instance()._calculateMinHeightOptionList = jest.fn(() => minHeight);

        // Action
        const result = shallow(multipleSelectOptionList.instance().render());

        // Assert
        expect(result.props().className).toBe('multiple-select-options');
        expect(result.props().style).toEqual({ minHeight });
        expect(_renderOptionList).toHaveBeenCalled();
    });
});