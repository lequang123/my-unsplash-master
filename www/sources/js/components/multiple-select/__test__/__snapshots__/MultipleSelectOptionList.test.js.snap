// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleSelectOptionList should have correct snapshots 1`] = `ShallowWrapper {}`;
